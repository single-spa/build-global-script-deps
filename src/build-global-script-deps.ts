import { x } from "tar";
import fs from "node:fs/promises";

interface Options {
  globalScripts: string[];
  outputDir: string;
}

export async function buildGlobalScripts(options: Options) {
  const outputDir = options.outputDir.startsWith(".")
    ? options.outputDir
    : `./${options.outputDir}`;

  for (let globalScript of options.globalScripts) {
    console.log(globalScript);
    const packageName = globalScript.slice(0, globalScript.lastIndexOf("@"));
    const version = globalScript.slice(globalScript.lastIndexOf("@") + 1);
    const registryResponse = await fetch(
      `https://registry.npmjs.org/${packageName}`,
    );
    if (!registryResponse.ok) {
      throw Error(`Unable to find package '${packageName}' in npm registry`);
    }
    const registryJson = await registryResponse.json();
    const versionInfo = registryJson.versions[version];
    if (!versionInfo) {
      throw Error(
        `package '${packageName}' has not published version '${version}'`,
      );
    }

    const tarballResponse = await fetch(versionInfo.dist.tarball);

    if (!tarballResponse.ok) {
      throw Error(`Could not download tarball for package '${globalScript}'`);
    }

    await fs.mkdir(outputDir, { recursive: true });

    // To-do figure out how to pipe the fetch response into tar
    const contents = await tarballResponse.arrayBuffer();

    const tarballPath = `${outputDir}/tarball.tar`;

    await fs.writeFile(tarballPath, Buffer.from(contents), "utf-8");

    await x({
      f: tarballPath,
      cwd: outputDir,
    });

    const outputFolder = `${outputDir}/npm:${globalScript}`;

    // Create top-level folder for scoped packages
    const outputFolderParts = outputFolder.split("/");
    if (outputFolderParts.length > 2) {
      const pathParts = outputFolder.split("/");
      pathParts.pop();
      await fs.mkdir(pathParts.join("/"), { recursive: true });
    }

    await fs.rename(`${outputDir}/package`, outputFolder);

    await fs.unlink(tarballPath);
  }
}
