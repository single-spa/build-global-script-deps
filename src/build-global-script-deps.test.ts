import { describe, expect, it, jest } from "@jest/globals";
import { vol, fs } from "memfs";
import { buildGlobalScripts } from "./build-global-script-deps";
import path from "node:path";
import { beforeEach } from "node:test";

jest.mock("fs");
jest.mock("fs/promises");

describe(`build-global-script-deps`, () => {
  beforeEach(() => {
    vol.reset();
  });

  it(`writes folders to the file system`, async () => {
    await buildGlobalScripts({
      outputDir: "deps",
      globalScripts: [
        "import-map-overrides@4.2.0",
        "@single-spa/import-map-injector@2.0.0",
      ],
    });

    expect(readDirRelative("deps")).toMatchSnapshot();
  });
});

function readDirRelative(dir: string) {
  return fs
    .readdirSync(dir, { recursive: true })
    .map((file) => path.relative(process.cwd(), file))
    .sort();
}
