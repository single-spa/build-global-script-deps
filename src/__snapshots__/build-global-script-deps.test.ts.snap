// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build-global-script-deps writes folders to the file system 1`] = `
[
  "npm:@single-spa",
  "npm:@single-spa/import-map-injector@2.0.0",
  "npm:@single-spa/import-map-injector@2.0.0/LICENSE",
  "npm:@single-spa/import-map-injector@2.0.0/README.md",
  "npm:@single-spa/import-map-injector@2.0.0/lib",
  "npm:@single-spa/import-map-injector@2.0.0/lib/import-map-injector.js",
  "npm:@single-spa/import-map-injector@2.0.0/lib/import-map-injector.js.map",
  "npm:@single-spa/import-map-injector@2.0.0/package.json",
  "npm:@single-spa/import-map-injector@2.0.0/types",
  "npm:@single-spa/import-map-injector@2.0.0/types/import-map-injector.d.ts",
  "npm:import-map-overrides@4.2.0",
  "npm:import-map-overrides@4.2.0/LICENSE",
  "npm:import-map-overrides@4.2.0/README.md",
  "npm:import-map-overrides@4.2.0/dist",
  "npm:import-map-overrides@4.2.0/dist/import-map-overrides-api.js",
  "npm:import-map-overrides@4.2.0/dist/import-map-overrides-api.js.map",
  "npm:import-map-overrides@4.2.0/dist/import-map-overrides-server.js",
  "npm:import-map-overrides@4.2.0/dist/import-map-overrides-server.js.map",
  "npm:import-map-overrides@4.2.0/dist/import-map-overrides.js",
  "npm:import-map-overrides@4.2.0/dist/import-map-overrides.js.map",
  "npm:import-map-overrides@4.2.0/package.json",
]
`;
